### PR TITLE
fix(ts): adds function type definition for `ComponentOptions.middleware`

### PR DESCRIPTION
**Pull request by hartmut-co-uk**
Better TS support: Adds function typings to `ComponentOptions.middleware`.

## Types of changes
- [x] Bug fix (a non-breaking change which fixes an issue)
- [ ] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description
After discussion with @kevinmarrec added type defs for functions.

## Checklist:
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly. (PR: #)
- [x] I have added tests to cover my changes (if not applicable, please state why)
- [ ] All new and existing tests are passing.

### CODE REVIEW

File: packages/vue-app/types/vue.d.ts [hunk 1]
```diff
@@ -14,7 +14,7 @@ declare module "vue/types/options" {
     head?: MetaInfo | (() => MetaInfo);
     key?: string | ((to: Route) => string);
     layout?: string | ((ctx: Context) => string);
-    middleware?: string | string[];
+    middleware?: string | string[] | ((ctx: Context, cb: Function) => Promise<void>| void);
```
Comment (kevinmarrec) [inline packages/vue-app/types/vue.d.ts]: Missing space around `Promise<void>|` 😇

**Review by kevinmarrec — CHANGES_REQUESTED** [review]

File: packages/vue-app/types/vue.d.ts (hunk 1 above)
Comment (hartmut-co-uk) [reply packages/vue-app/types/vue.d.ts]: updated

**Review by kevinmarrec — CHANGES_REQUESTED** [review]
@hartmut-co-uk I think that the 2 last tests can be omitted 🤔

File: test/types/index.ts [hunk 2]
```diff
@@ -53,6 +53,20 @@ options.layout = (context) => 'foo'
...
 
+options.middleware = (ctx) => {}
+options.middleware = (ctx, cb) => { cb() }
+options.middleware = async (ctx) => {}
+
+options.middleware = async (ctx) => {
+  let res = await Promise.resolve('bar')
+}
+
+options.middleware = (ctx, cb) => {
```
[the remark this answers is not in the document]
Comment (kevinmarrec) [reply test/types/index.ts]: Or at least we can replace it by 
```js
options.middleware = async (ctx, cb) => {}
```
to test the type
`(ctx: Context, cb: Function) => Promise<void>`
But I don't think this way will be used 😄

[the remark this answers is not in the document]
Comment (kevinmarrec) [reply test/types/index.ts]: The `async` way should not be used with a callback

**Comment by hartmut-co-uk** [comment]
Thx @kevinmarrec
Note: after further discussion on nuxt Discord updated the PR with @kevinmarrec's final suggestion to introduce `Middleware` type + more elegant tests

**Review by kevinmarrec — APPROVED** [review]
Everything is perfect to me.

PR is merge-ready !

**Review by galvez — APPROVED** [review]
LGTM!

**Pull request merged**